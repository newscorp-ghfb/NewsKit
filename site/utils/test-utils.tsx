/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {
  renderToFragmentWithThemeFactory,
  renderWithThemeFactory,
} from 'newskit/test/test-utils';
import {docsThemeLight} from '../theme/doc-theme';

export const renderWithTheme = renderWithThemeFactory(docsThemeLight);

export const renderToFragmentWithTheme: any = renderToFragmentWithThemeFactory(
  docsThemeLight,
);

export const mockNewsKitComponents = (...components: string[]) => () => ({
  ...(jest.requireActual('newskit') as any),
  ...components.reduce((acc, comp) => {
    acc[comp] = ({children, ...props}: any) => (
      <div
        data-comp={`NewsKit ${comp}`}
        data-testid={comp}
        data-props={JSON.stringify(props, null, 2)}
      >
        {children}
      </div>
    );
    return acc;
  }, {} as Record<string, Function>),
});
