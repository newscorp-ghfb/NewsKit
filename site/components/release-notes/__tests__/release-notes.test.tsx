import {renderWithTheme} from 'newskit/test/test-utils';
import React from 'react';
import {Matcher} from '@testing-library/react';
import ReleaseNotes from '../release-notes';
import '@testing-library/jest-dom';

jest.mock('react-markdown', () => ({
  __esModule: true,
  default: ({children}: {children: React.ReactNode}) => <>{children}</>,
}));

const COMMENT =
  '<!-- Release notes generated using configuration in .github/release.yml at main -->';
const PROFILE = '@mstuartf';
const PR_LINK = 'https://github.com/newscorp-ghfb/newskit/pull/239';
const COMPARE_LINK =
  'https://github.com/newscorp-ghfb/newskit/compare/v5.6.0...trigger-release@5.7.0';
const TICKET_ID = 'PPDSC-2121';

const BODY = `
${COMMENT}\\r\\n
\\r\\n
## What's Changed\\r\\n
### New Features 🎉\\r\\n
* feat(PPDSC-2151): Add Outline Style Presets by @tbradbury in https://github.com/newscorp-ghfb/newskit/pull/214\\r\\n
* feat(${TICKET_ID}): \`<Popover/>\` header close by ${PROFILE} in ${PR_LINK}\\r\\n
### Bug Fixes 🐛\\r\\n
* fix(PPDSC-2164): assistive text by @mstuartf in https://github.com/newscorp-ghfb/newskit/pull/249\\r\\n
### Other Changes 🧱\\r\\n
* ci(PPDSC-2292): attempt fix of develoment env by @jps in https://github.com/newscorp-ghfb/newskit/pull/285\\r\\n
\\r\\n
## New Contributors\\r\\n
* @agagotowiec made their first contribution in https://github.com/newscorp-ghfb/newskit/pull/238\\r\\n
\\r\\n
**Full Changelog**: ${COMPARE_LINK}
`;

const partialQueryByText = (
  queryByText: (text: Matcher) => HTMLElement | null,
  partialContents: string,
) => queryByText(content => content.includes(partialContents));

describe('ReleaseNotes', () => {
  test('renders correctly', () => {
    const {asFragment} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(asFragment()).toMatchSnapshot();
  });
  test('removes comment', () => {
    const {queryByText} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(partialQueryByText(queryByText, COMMENT)).not.toBeInTheDocument();
  });
  test('adds PR links', () => {
    const {queryByText} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(
      partialQueryByText(
        queryByText,
        `[#${PR_LINK.split('/').slice(-1)[0]}](${PR_LINK})`,
      ),
    ).toBeInTheDocument();
  });
  test('adds ticket links', () => {
    const {queryByText} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(
      partialQueryByText(
        queryByText,
        `[${TICKET_ID}](https://nidigitalsolutions.jira.com/browse/${TICKET_ID})`,
      ),
    ).toBeInTheDocument();
  });
  test('adds GitHub profile links', () => {
    const {queryByText} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(
      partialQueryByText(
        queryByText,
        `[${PROFILE}](https://github.com/${PROFILE.replace('@', '')})`,
      ),
    ).toBeInTheDocument();
  });
  test('corrects and hyperlinks compare releases links', () => {
    const {queryByText} = renderWithTheme(ReleaseNotes, {body: BODY});
    expect(
      partialQueryByText(queryByText, COMPARE_LINK),
    ).not.toBeInTheDocument();
    const fixed = COMPARE_LINK.replace('trigger-release@', 'v');
    expect(
      partialQueryByText(queryByText, `[${fixed}](${fixed})`),
    ).toBeInTheDocument();
  });
});
