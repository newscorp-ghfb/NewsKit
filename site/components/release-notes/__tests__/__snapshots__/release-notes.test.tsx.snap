// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReleaseNotes renders code snippets correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin-block-start: 32px;
}

.emotion-1 {
  margin: 0;
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-2 {
  background-color: rgba(27, 31, 35, 0.05);
  color: #6A6A6A;
  border-radius: 3px;
  font-size: 85%;
  margin: 0;
  padding: 0.2em 0.4em;
  font-family: monospace;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <h3>
        Emotion packages as peer dependencies\\r\\nStarting from version 6, newskit does not include 
        <code
          class="emotion-2"
        >
          @emotion/styled
        </code>
         & 
        <code
          class="emotion-2"
        >
          @emotion/react
        </code>
         libraries, which means you need to install these separately:\\r\\n
        <code
          class="emotion-2"
        >
          \\r\\n// npm\\r\\nnpm install newskit@6.0.0 @emotion/react @emotion/styled\\r\\n// or yarn\\r\\nyarn add newskit@6.0.0 @emotion/react @emotion/styled\\r\\n
        </code>
        \\r\\n---\\r\\n
      </h3>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ReleaseNotes renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin-block-start: 32px;
}

.emotion-1 {
  margin: 0;
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-1::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-1::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-2 {
  display: inline-block;
  color: #3358CC;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-2 {
    transition-property: color,fill;
    transition-duration: 200ms,200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1),cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-2 {
    transition-property: color,fill;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1),cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-2 svg {
  fill: #3358CC;
}

.emotion-2:hover:not(:disabled) {
  color: #254CAC;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-2:hover:not(:disabled) svg {
  fill: #254CAC;
}

.emotion-2:active:not(:disabled) {
  color: #14328D;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-2:active:not(:disabled) svg {
  fill: #14328D;
}

.emotion-2:focus-visible:not(:disabled) {
  outline-color: #3768FB;
  outline-style: solid;
  outline-width: 2px;
  outline-offset: 2px;
}

@media not all and (min-resolution: 0.001dpcm) {
  @supports (-webkit-appearance: none) and (stroke-color: transparent) {
    .emotion-2:focus-visible:not(:disabled) {
      outline-style: auto;
    }
  }
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-left: calc(-4px/2);
  margin-right: calc(-4px/2);
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  margin-left: calc(4px/2);
  margin-right: calc(4px/2);
}

.emotion-5 {
  margin: 0;
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
  display: inline-block;
  display: block;
}

.emotion-5::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-5::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

.emotion-7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
  fill: #3B3B3B;
  vertical-align: unset;
  display: inline-block;
}

@media screen and (prefers-reduced-motion: no-preference) {
  .emotion-7 {
    transition-property: fill;
    transition-duration: 200ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

@media screen and (prefers-reduced-motion: reduce) {
  .emotion-7 {
    transition-property: fill;
    transition-duration: 0ms;
    transition-timing-function: cubic-bezier(0, 0, .5, 1);
  }
}

.emotion-7.emotion-7 {
  width: 16px;
  height: 16px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <h3>
        New Features ðŸŽ‰
      </h3>
      

      <ul>
        

        <li>
          feat(
          <a
            class="emotion-2"
            href="https://nidigitalsolutions.jira.com/browse/PPDSC-2151"
          >
            <span
              class="emotion-3"
            >
              <span
                class="emotion-4"
              >
                <span
                  class="emotion-5"
                >
                  PPDSC-2151
                </span>
              </span>
              <span
                class="emotion-4"
              >
                <svg
                  class="emotion-7 emotion-8"
                  fill="currentColor"
                  focusable="false"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    External link
                  </title>
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 19H5V5h7V3H5a2 2 0 00-2 2v14a2 2 0 002 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                  />
                </svg>
              </span>
            </span>
          </a>
          ): Add Outline Style Presets by 
          <a
            class="emotion-2"
            href="https://github.com/tbradbury"
          >
            <span
              class="emotion-3"
            >
              <span
                class="emotion-4"
              >
                <span
                  class="emotion-5"
                >
                  @tbradbury
                </span>
              </span>
              <span
                class="emotion-4"
              >
                <svg
                  class="emotion-7 emotion-8"
                  fill="currentColor"
                  focusable="false"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    External link
                  </title>
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 19H5V5h7V3H5a2 2 0 00-2 2v14a2 2 0 002 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                  />
                </svg>
              </span>
            </span>
          </a>
           in 
          <a
            class="emotion-2"
            href="https://github.com/newscorp-ghfb/newskit/pull/214"
          >
            <span
              class="emotion-3"
            >
              <span
                class="emotion-4"
              >
                <span
                  class="emotion-5"
                >
                  #214
                </span>
              </span>
              <span
                class="emotion-4"
              >
                <svg
                  class="emotion-7 emotion-8"
                  fill="currentColor"
                  focusable="false"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title>
                    External link
                  </title>
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                  <path
                    d="M19 19H5V5h7V3H5a2 2 0 00-2 2v14a2 2 0 002 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                  />
                </svg>
              </span>
            </span>
          </a>
        </li>
        

      </ul>
    </div>
  </div>
</DocumentFragment>
`;
